crates/api/src/chronicle_graphql/mod.rs

```diff
@@ -373,37 +374,134 @@ where
     }
 }
 
+#[derive(Clone, Debug, Serialize)]
+pub enum AccessType {
+    QUERY,
+    MUTATE,
+    SUBSCRIBE,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct AuthGuard {
+    access_type: AccessType,
+    method_name: String,
+}
+
+impl AuthGuard {
+    pub fn new(access_type: AccessType, method_name: impl ToString) -> Self {
+        AuthGuard {
+            access_type,
+            method_name: method_name.to_string(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl async_graphql::Guard for AuthGuard {
+    async fn check(&self, ctx: &Context<'_>) -> async_graphql::Result<()> {
+        if let (Some(identity), Some(opa_executor)) = (
+            ctx.data_opt::<AuthId>(),
+            ctx.data_opt::<Arc<Mutex<WasmtimeOpaExecutor>>>(),
+        ) {
+            let context = serde_json::to_value(self)?;
+            if opa_executor
+                .lock()
+                .await
+                .evaluate(identity, &context)
+                .await?
+            {
+                Ok(())
+            } else {
+                Err("violation of policy rules".into())
```

Comment:
Can we log something to info with more detail here (identity + context), and #[instrument] the guard method. The error ends up in the client, so keeping that mysterious is good but operationally we need more details here